AbstractSQLCompiler.ts

```diff
@@ -359,15 +367,164 @@ const compileSchema = (abstractSqlModel: AbstractSqlModel, engine: Engines, ifNo
 	}
 }
 
+const genSplit = <T>(src:Array<T>, dst: Array<T>, matchFn: MatchFn<T>): Split<T> => {
+	let mod: Array<Pair<T>> = []
+	return _.reduce(src, (acc, value) => {
+		const match = matchFn(acc.ins, value)
+		if (_.isUndefined(match)) {
+			return acc
+		} else {
+			acc.ins = _.without(acc.ins, match)
+			acc.del = _.without(acc.del, value)
+			acc.mod.push( { src: value, dst: match } )
+			return acc
+		} 
+	}
+	, { ins: dst, del: src, mod: mod } )
+}
+//  genDiff expects src and dst arrays formatted as described in
+//  genSplit, and three functions to be called in the case of
+//  insertion, deletion or modification.
+//  All these functions are treated as reducers, their first argument
+//  will be an accumulator of all the changes up to that point and
+//  they are expected to add their changes to it and return it.
+//  The second argument will be an array of [key, value] for
+//  insertions and deletions and [key, pair] for the modifications.
+//  The match function describes how to find the corresponding T inside a T[].
+//  The result of the match function will be either the element that should be matched to 
+//  the argument or undefined if no such element is found.
+const genDiff = <T>(insFn: ReduceFn<T>, delFn: ReduceFn<T>, modFn: ReduceFn<Pair<T>>, matchFn: MatchFn<T>, src: Array<T>, dst: Array<T>) => {
+	let split = genSplit(src, dst, matchFn)
+	return  _.flatten(_.reduce(split.ins, insFn, 
```

Comment:
I think all this code could be made much nice with either
```typescript
_.flatMap(split.mod, modFn)
.concat(_.flatMap(split.del, delFn))
.concat(_.flatMap(split.ins, insFn))
```
and then `return []` to include nothing, and `return [a, b, c]` to return multiple.  It avoids all the need for mutation and overall should end up much nicer imo.  The alternative if you only ever need to return 0 or 1 results (never > 1) is:
```typescript
_.map(split.mod, modFn)
.concat(_.map(split.del, delFn))
.concat(_.map(split.ins, insFn))
.reject(_.isNil)
```
and `return` for nothing, `return a` for something